### PR TITLE
Speed up checksum calculation in io.fits (alternative)

### LINKED ISSUE #16372: Faster checksum implementation
### What is the problem this feature will solve?

Astro.py currently takes several seconds to add a datasum to large fits files. In my tests using a 1.1GB fits file, where most of the data is in the primary HDU, it took ~2.5s to perform the following:

```python
hdul = fits.open('inversion_output/scan1807/inv_res_pre.fits')
hdul[0].add_datasum()
hdul[0].header
```

### Describe the desired outcome

While implementing this in Haskell, I used the c code in the FITS 4.0 spec, and adding a datasum to the same HDU was about 10x faster. 

Code is included here. There are small differences from the code in the FITS 4.0 spec, namely returning the resulting sum instead of passing in a pointer. 

```c

unsigned int checksum (
  unsigned char *buf,
  int length)
{

  unsigned int sum32 = 0;
  unsigned int hi, lo, hicarry, locarry, i;

  // Accumulate the sum of the high-order 16 bits and the */
  // low-order 16 bits of each 32-bit word, separately. */
  // The first byte in each pair is the most significant. */
  // This algorithm works on both big and little endian machines.
  hi = (sum32 >> 16);
  lo = sum32 & 0xFFFF;
  for (i=0; i < length; i+=4) {
    hi += ((buf[i] << 8) + buf[i+1]);
    lo += ((buf[i+2] << 8) + buf[i+3]);
  }

  // fold carry bits from each 16 bit sum into the other sum
  hicarry = hi >> 16;
  locarry = lo >> 16;

  while (hicarry || locarry) {
    hi = (hi & 0xFFFF) + locarry;
    lo = (lo & 0xFFFF) + hicarry;
    hicarry = hi >> 16;
    locarry = lo >> 16;
  }

  // Concatenate the high and low parts to form the full 32-bit checksum
  return (hi << 16) + lo;
}
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mhvk**
This is an alternative to #17205, which optimizes the checksum calculation within numpy itself, by realizing that the routine taken from FITS is just a very complicated way to add unsigned int 32 together, written perhaps in a time that unsigned int 64 did not exist.

Timing with this PR is essentially identical to that of #17205, but the code is much simpler (though still need a loop to ensure that also for files larger than 16GB, the sum cannot overflow).

Note that this includes the commit from #17205 with the changelog entry as well as the one that changes a `.flatten()` to `.ravel()` to avoid a copy. The latter is definitely OK, since we only read the data. Asking not to squash commits so that proper credit is given.

p.s. Not running benchmark, since for #17205, it didn't show anything, so clearly there is no relevant benchmark. Feel free to push one to this PR.

Fixes #16372

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Thanks for the ping. At first glance this is indeed much simpler, and I'd favor any solution that doesn't involve adding native code if performances are comparable. I'll review in detail shortly !

**Comment by pllim** [comment]
oldestdeps failure looks related

`TypeError: ufunc 'right_shift' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''`

**Comment by pllim** [comment]
p.s. Might wanna apply "Extra CI" label too when that failure is fixed.

**Comment by mhvk** [comment]
OK, should be fixed. Good idea to also run Extra CI so we get coverage of bigendian.

File: astropy/io/fits/hdu/base.py [hunk 1]
```diff
@@ -1464,55 +1464,23 @@ def _compute_checksum(self, data, sum32=0):
...
+        # was there before.
+        s = int(np.uint32(sum32))
+        for piece in np.split(data, range(blocklen, len(data), blocklen)):
+            if extra := piece.nbytes % 4:
+                last = bytes(piece[-extra:]) + b"\00" * (4 - extra)
+                s += int.from_bytes(last, byteorder="big", signed=False)
+                piece = piece[:-extra]
+            s += int(piece.view(">u4").sum(dtype="u8"))
+            while hi := (s >> 32):
+                s = (s & 0xFFFFFFFF) + hi
```
Comment (neutrinoceros) [inline astropy/io/fits/hdu/base.py:1482]: If I understand correctly, what's happening here is that we're compressing information from a possibly unlimited sized `int` down to a `uint32`, but it looks to me that information is still lost if `s` initially doesn't fit in a `uint64` (specifically the loss happens on the masking operation `s & 0xFFFFFFFF`). This may never be a problem in practice, but I'm trying to make sure I'm getting the hang of it, and if I'm correct, I think this aspect should be explained in a comment too.

**Review by neutrinoceros — COMMENTED** [review]
I find the new implementation much clearer, in the sense that I could not understand what the previous implementation was doing in a reasonable amount of time, and I trust that the results are equivalent since tests pass, *but* I would still recommend adding a couple comments here and there because it is *still* somewhat hard to follow in my humble opinion.
I want to formally approve once my questions are answered.

File: astropy/io/fits/hdu/base.py [hunk 2]
```diff
@@ -1464,55 +1464,23 @@ def _compute_checksum(self, data, sum32=0):
...
+        # The cast to uint32 is not needed by the tests, and seems odd, since
+        # we explicitly use higher bits otherwise. Kept regardless since it
+        # was there before.
+        s = int(np.uint32(sum32))
+        for piece in np.split(data, range(blocklen, len(data), blocklen)):
+            if extra := piece.nbytes % 4:
+                last = bytes(piece[-extra:]) + b"\00" * (4 - extra)
+                s += int.from_bytes(last, byteorder="big", signed=False)
+                piece = piece[:-extra]
+            s += int(piece.view(">u4").sum(dtype="u8"))
```
Comment (astrofrog) [inline astropy/io/fits/hdu/base.py:1480]: Just curious, would there be any performance benefit to doing the view and sum in a vectorised way or would it be tricky because it would require padding the array first to be a multiple of the chunk size?

**Review by mhvk — COMMENTED** [review]
Thanks for the good comments. Answers inline. I've pushed an updated version (mostly just clarifications).

**Comment by mhvk** [comment]
Failures on exotics are unrelated (they passed before with the same code). Something seems to have gone wrong with the basic setup:
```
  #3 [internal] load metadata for docker.io/s390x/ubuntu:rolling
  #3 ERROR: no match for platform in manifest: not found
  ------
   > [internal] load metadata for docker.io/s390x/ubuntu:rolling:
  ------
  Dockerfile.s390x.ubuntu_rolling:1
  --------------------
     1 | >>> FROM s390x/ubuntu:rolling
     2 |     
     3 |     COPY ./run-on-arch-install.sh /root/run-on-arch-install.sh
  --------------------
  ERROR: failed to solve: s390x/ubuntu:rolling: failed to resolve source metadata for docker.io/s390x/ubuntu:rolling: no match for platform in manifest: not found
```

**Comment by neutrinoceros** [comment]
Interesting. Indeed the failure also shows up on `main`, though it means `Extra CI` is currently useless, which comes at a rather bad time. I'll open an issue and see what I can do about it.

File: astropy/io/fits/hdu/base.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/io/fits/hdu/base.py:1483]: > hi carries all the information of the higher bits. If those are more than 32 bits (or the sum is over 32 bits), then after the sum, s will have more than 32 bits, and hence the while runs one more time.

yes, but my comment is about `s` overflowing a `uint64` (not `uint32`). Again, very unlikely that this is ever a problem in practice, but am I right that information beyond the 64th bit would be lost ?

File: astropy/io/fits/hdu/base.py [hunk 3]
```diff
@@ -1464,55 +1464,23 @@ def _compute_checksum(self, data, sum32=0):
...
-        if data.nbytes % 2:
-            last = data[-1]
-            data = data[:-1]
-        else:
-            last = np.uint32(0)
-
-        data = data.view(">u2")
-
-        hi = sum32 >> u16
-        lo = sum32 & uFFFF
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/io/fits/hdu/base.py:1497]: Oh, I see it now; I read this wrong (I'm really not used to hexadecimal) and thought that each `F` represented 8 bits when it's really 4, hence `& uFFFF` is only masking 16 bits instead of 32 as I assumed.
Thanks for the explanation though, it's really useful !

File: astropy/io/fits/hdu/image.py [hunk 4]
```diff
@@ -945,7 +945,7 @@ def _calculate_datasum(self):
             else:
                 byteswapped = False
 
-            cs = self._compute_checksum(d.flatten().view(np.uint8))
```
Comment (neutrinoceros) [inline astropy/io/fits/hdu/image.py:948]: just to be clear, the difference between `ndarray.flatten` and `ndarray.ravel` is that the former returns a copy, while the latter doesn't, correct ?

File: astropy/io/fits/hdu/base.py (hunk 1 above)
Comment (mhvk) [reply astropy/io/fits/hdu/base.py:1483]: Yes, it we would overflow 64 bits, that information would be lost, but given the blocklen, that cannot possibly happen, even if every `uint32=0xffffffff` (i.e., `blocklen * 0xffffffff < 0xffffffffffffffff`). See comment about blocklen.

File: astropy/io/fits/hdu/image.py (hunk 4 above)
Comment (mhvk) [reply astropy/io/fits/hdu/image.py:948]: Yes, `ravel()` returns a view (unless it is not possible because the strides are not C aligned, then it makes a copy), while `flatten()` always makes a copy.

**Comment by mhvk** [comment]
Rebased so Extra CI can run properly.

**Comment by pllim** [comment]
Benchmark says "nothing significantly changed" but CI is all green.

**Comment by mhvk** [comment]
I think we just don't have a benchmark for this... (I don't think I've got the time to remedy that now, unfortunately).

**Review by saimn — APPROVED** [review]
Nice alternative, perf is similar to the cython version so works for me. Thanks @mhvk !
